spec.md

```diff
@@ -2110,6 +2215,72 @@ Condition | gRPC Code | Description | Recovery Behavior
 | Call not implemented | 12 UNIMPLEMENTED | NodeGetInfo call is not implemented by the plugin or disabled in the Plugin's current mode of operation. | Caller MUST NOT retry. Caller MAY call `ControllerGetCapabilities` or `NodeGetCapabilities` to discover Plugin capabilities. |
 
 
+#### `NodeExpandVolume`
+
+A Node Plugin MUST implement this RPC call if it has `EXPAND_VOLUME` node capability.
+This RPC call allows CO to expand volume on the node.
+
+If plugin has `PUBLISH_UNPUBLISH_VOLUME` controller capability:
+* `NodeExpandVolume` MUST be called after calling `ControllerPublishVolume`.
+* If plugin has `STAGE_UNSTAGE_VOLUME` node capability then `NodeExpandVolume` MUST be called after `NodeStageVolume`.
+* `NodeExpandVolume` however MAY be called before or after `NodePublishVolume`.
+
+If plugin does not have `PUBLISH_UNPUBLISH_VOLUME` controller capability:
+* `NodeExpandVolume` MUST be called after `NodeStageVolume` if plugin has `STAGE_UNSTAGE_VOLUME` node capability.
+* `NodeExpandVolume` however MAY be called before or after `NodePublishVolume`.
+
+`NodeExpandVolume` expects volume to be available on specified `volume_path` and `NodeExpandVolume` itself will not implicitly call `NodeStageVolume` or `NodePublishVolume`.
+
+`NodeExpandVolume` may also be called to perform online expansion of in-use Volume when volume supports `ONLINE` `OperationMode` for expansion.
+A plugin may return `FAILED_PRECONDITION` error if underlying volume can not be expanded when in-use.
+
+Examples:
+* Offline Volume Expansion:
+  Given a ElasticSearch process that runs on EBS volume and needs more space.
+  - The administrator takes the elasticsearch server offline and request more space for volume from the CO.
+  - The CO in turn first makes `ControllerExpandVolume` RPC call which results in requesting more space from AWS-EBS for volume-id that was being used by ElasticSearch.
+  - Once `ControllerExpandVolume` finishes the CO will inform administrator about it and administrator will resume the ElasticSearch workload.
+  - On node where ElasticSearch workload is scheduled when workload starts after `NodeStageVolume` the CO calls `NodeExpandVolume`.
```

Comment:
There are two aspects of it right? 

a. You are using a volume type that does not support online expansion of already controller published volumes. But the file system can be expanded online (ex. Azure Disk + xfs)
b. The SP supports online expansion but the file system you are using can not be expanded when mounted. (ex. GCE-PD + ext2)

What you are pointing out mostly applies to option *b*.  I think this brings up another question - two operation modes `ONLINE` and `OFFLINE` might not be sufficient to describe volume capabilities since controller and node expansion properties could be different.

Reply:
I am thinking of updating volume capability:

```
  message Expandable {
    enum NodeOperationMode {
      UNKNOWN = 0;

      // Can be expanded when in-use by a workload.
      ONLINE = 1;

      // Can only be expanded when not in-use.
      OFFLINE = 2;
    }
    enum ControllerOperationMode {
      UNKNOWN = 0;

      // Can be expanded when in-use by a workload.
      ONLINE = 1;

      // Can only be expanded when not in-use.
      OFFLINE = 2;
    }
  }
```

So as we can cover expansion capabilities of node and controller separately. Supporting resizing of unmounted volumes on the node will also require expanding `NodeExpandVolume` RPC call to take same/similar parameters as `NodeStageVolume`. @chhsia0 @jieyu what you think? 

Another option is:

Declare `NodeExpandVolume` as a operation that allows volume expansion for only mounted/online  volumes. Any volume that only has `offline` capability (Including Azure Disk+XFS from above example) MUST do fs resizing on the node as part of node stage/publish process.  This would simplify the workflow at the cost of some extra work on plugin part. Such volumes will probably return `false` as `fs_resize_required` because filesystem resizing is implicit on stage/publish on the node.

Reply:
@gnufied hum, both solutions do not sound ideal :( OK, now I sort of feel that maybe we don't support offline as a start given the complexity involved in the spec? Given that most fs already supports online expansion?

Reply:
updated the proposal as per our discussion.